pkg/antctl/raw/supportbundle/command.go

```diff
@@ -692,9 +702,151 @@ func processResults(resultMap map[string]error, dir string) error {
 		err = writeFailedNodes(dir, failedNodes)
 	}
 
+	// download logs from kubernetes api
+	if failedNodes != nil {
+		err = downloadAgentBundleFromKubernetes(ctx, antreaClientset, k8sClient, failedNodes, dir)
+		if err != nil {
+			fmt.Println("Failed to download agent bundle from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+	if controllerFail {
+		err = downloadControlleBundleFromKubernetes(ctx, antreaClientset, k8sClient, dir)
+		if err != nil {
+			fmt.Println("Failed to download controller bundle from kubernetes api: " + err.Error())
+		} else {
+			allFailed = false
+		}
+	}
+
 	if allFailed {
 		return fmt.Errorf("no data was collected: %s", genErrorMsg(resultMap))
 	} else {
 		return err
 	}
 }
+
+func downloadControlleBundleFromKubernetes(ctx context.Context, antreaClientset antrea.Interface, k8sClient kubernetes.Interface, dir string) error {
+	var errors []error
+	tmpDir, err := afero.TempDir(defaultFS, "", "bundle_tmp_")
+	if err != nil {
+		return err
+	}
+	defer defaultFS.RemoveAll(tmpDir)
+
+	controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(ctx, "antrea-controller", metav1.GetOptions{})
+	if err == nil {
+		data, err := yaml.Marshal(controllerInfo)
+		if err == nil {
+			err = afero.WriteFile(defaultFS, filepath.Join(tmpDir, "controllerinfo"), data, 0644)
+			errors = append(errors, err)
+		}
+	}
```

Comment:
I think the code is more readable and less error-prone if you do something like this:
```go
if err := func() {
        controllerInfo, err := antreaClientset.CrdV1beta1().AntreaControllerInfos().Get(ctx, "antrea-controller", metav1.GetOptions{})
        if err != nil {
                return err
        }
        data, err := yaml.Marshal(controllerInfo)
        if err != nil {
                return err
        }
        if err := afero.WriteFile(defaultFS, filepath.Join(tmpDir, "controllerinfo"), data, 0644); err != nil {
                return err
        }
        return nil
}(); err != nil {
        errors = append(errors, err)
}
```

while it is a bit more verbose, it follows a more usual pattern

Reply:
Do we need to support the case where multiple controller pod running? I know it's not recommended and users should not do this, but i sure can just scale up the deployment. In this case,  should we identify the pod by podRef in controllerinfo?

Reply:
Yes personally I think it's better to rely on the `podRef` field in the `AntreaControllerInfo`, instead of doing a `List` API operation. During upgrade, the `List` may return 2 entries.

Reply:
updated